lib/runtime/printk.c

```diff
@@ -219,17 +178,45 @@ struct out_context {
 	int count;
 };
 
-static int char_out(int c, struct out_context *ctx)
+
+#ifdef CONFIG_USERSPACE
+K_SEM_DEFINE(printk_sem, 1, 1);
+
+struct printk_buffer {
+	int count;
+	char buf[CONFIG_PRINTK_BUFFER_SIZE];
+};
+
+static struct printk_buffer pbuf;
+
+static void char_out(int c, struct out_context *ctx)
```

Comment:
This is a behavior change: before, you could count that printk() would flush the output every time it was called.

Now it expects either the buffer to be filled, or a new line character to appear in the stream.  Programs that were written with a series of printk() calls won't produce the same output, and depending on the output, won't produce any output at all.

The buffer should either be flushed after _vprintk() is done, or k_str_out() may be called with just 1 character at a time until printk() is modified in a way that the syscall is called for every chunk to be printed.

Reply:
i can add the buffer flush for the userspace=y case, seems simplest.